### DISCUSSION TITLE
fpv.py failed

### DISCUSSION THREAD

**Discussion by noahzn** [bitcraze/discussions#432] [Q&A] [answered]
I'm testing with the script fpv.py. I connect to the wifi SSID "wifi steaming example" and I use `URI = uri_helper.uri_from_env(default='tcp://192.168.4.1:5000')`. The drone flies for about 1 second and then stops. I can only see one image in the Qt window. The error is:

`Exception while reading transport, link probably closed?`
`[WinError 10054] An existing connection was forcibly closed by the remote host
 File "cflib\cpx\__init__.py", line 143, in run
    packet = self._transport.readPacket()
  File "cflib\cpx\transports.py", line 64, in readPacket
    size = struct.unpack('H', self._readData(2))[0]
  File "cflib\cpx\transports.py", line 60, in _readData
    data.extend(self._socket.recv(size-len(data)))
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host`

I'm using Windows 10. The `opencv-viewer.py` using WIFI works well.

I guess maybe it's a problem with my updating the firmware using the cfclient bootloader? What's the difference between "Flashing Wifi Example using `aideck_gap8_wifi_img_streamer_with_ap.bin` (https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/)" and "Configuration Wifi example (https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/test-functions/wifi-streamer/)"

update 26-11-2022:
I tried many times, and it succeeded three times. Most of the time, it took off for about one second, and then the connection was lost. And I noticed then the crazyflie restared automatically. 

update 26-11-2022 night:
I reinstalled the firmware 2022.05, and I also had the not taking-off problem. Then, I installed the firmware 2022.09, and flashed the gap8 bootloader. After that, I tried to flash the wifi_img_streamer_with_ap.bin, but it hangs at 6%.... I need help, please.

**Comment by knmcguire** [comment, marked as answer]
Hi! There seem to be some issues with the current FPV example as is unfortunately. So currently this is what works  for sure at the moment:

- Install the latest release (2022.09) on both the AIDeck and the crazyflie via the instructions of these: https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/
- Do not configure and reflash the Crazyflie for wifi (so just leave the current as is), just connect to the 'Wifi streaming example'
- On the cfclient when you connect to the crazyflie, you can see the AIDeck name in the console: `CPX: ESP32: I (800) DISCOVERY: Hostname is aideck-AABBCC` (The AABBCC is unique to your aideck)
- In the FPV.py file, replace the URI with `tcp://aideck-AABBCC.local:5000`

Unfortunately, it seems that reflashing the GAP8 bootloader seemed to have broken or bricked it somehow... but since this is a different topic please keep an eye on this discussion:  https://github.com/orgs/bitcraze/discussions/442. Once that has been fixed you can try what I've proposed above and report back.

**Reply by noahzn** [reply]
Thanks for your reply! I reinstalled the latest release, but sometimes the aideck wifi doesn't appear in my laptop's wifi list. 
`CPX: ESP32: I (524) DISCOVERY: Hostname is aideck-B0A3E0`, so I used `URI = uri_helper.uri_from_env(default='tcp://aideck-B0A3E0.local:5000')` or `URI ='tcp://aideck-B0A3E0.local:5000'`,  but got error `socket.gaierror: [Errno 11001] getaddrinfo failed`.

**Reply by knmcguire** [reply]
Could you copy the entire console output of the cfclient and paste it here?

**Reply by noahzn** [reply]
```
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:5c4261e5a9ff (2022.09) CLEAN
SYS: I am 0x20373357594B5007004B002C and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcFlow2 for deck 0
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
DECK_CORE: Calling INIT on driver bcAI for deck 1
CPX-EXT-ROUTER: Waiting for CPX External router initialization
CPX-EXT-ROUTER: CPX External router initialized
AIDECK: Not setting up WiFi
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2022.09
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
DECK_CORE: Deck 1 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 11712 bytes
CPX: ESP32: I (524) SYS: Initialized
CPX: ESP32: I (524) DISCOVERY: Hostname is aideck-B0A3E0
CPX: ESP32: I (524) DISCOVERY: MAC (STA) is 8:3a:f2:b0:a3:e0
CPX: GAP8: -- WiFi image streamer example --
STAB: Ready to fly.
CPX: GAP8: Setting up WiFi AP
```

**Reply by knmcguire** [reply]
Hmm oke that seems to be exactly the same as mine.

How about the cflib version? 

```
pip freeze | findstr cflib
```

**Reply by noahzn** [reply]
cflib==0.1.20.1
cfclient==2022.9

**Reply by knmcguire** [reply]
Ah there is a difference there, my cflib is from the latest development branch. You can try to install from source via [these instructions](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/installation/install/). It is possible to easily revert that with` pip3 uninstall cflib` and `pip3 install cflib`

Are you at least able to connect and see footage? If you disconnect the motors to prevent it from flying or comment out this line https://github.com/bitcraze/crazyflie-lib-python/blob/8d82a769952d1b49de0d6779a2b59ba94036151a/examples/aideck/fpv.py#L184, then at least you can check the video without it flying

**Reply by noahzn** [reply]
Now the cflib version is `-e git+https://github.com/bitcraze/crazyflie-lib-python.git@8d82a769952d1b49de0d6779a2b59ba94036151a#egg=cflib`

However, when I use `URI = uri_helper.uri_from_env(default='tcp://aideck-B0A3E0.local:5000')`, it cannot connect successfully, and shows: ` File "d:\project\crazyflie-lib-python\cflib\crazyflie\__init__.py", line 232, in open_link
    link_uri, self._link_quality_cb, self._link_error_cb)`
  `File "d:\project\crazyflie-lib-python\cflib\crtp\__init__.py", line 99, in get_link_driver
    instance.connect(uri, link_quality_callback, link_error_callback)`
  `File "d:\project\crazyflie-lib-python\cflib\crtp\tcpdriver.py", line 60, in connect
    self.cpx = CPX(SocketTransport(parse.hostname, parse.port))`
  `File "d:\project\crazyflie-lib-python\cflib\cpx\transports.py", line 38, in __init__
    self.connect()`
  `File "d:\project\crazyflie-lib-python\cflib\cpx\transports.py", line 43, in connect
    self._socket.connect((self._host, self._port))`
`socket.gaierror: [Errno 11001] getaddrinfo failed
Could not connect to Crazyflie`

If I use `URI = uri_helper.uri_from_env(default='tcp://192.168.4.1:5000')`, I can see the video stream for seconds this time, but then it froze, and errors appeared 

`Exception while reading transport, link probably closed?
[WinError 10054] An existing connection was forcibly closed by the remote host
Traceback (most recent call last):
  File "d:\project\crazyflie-lib-python\cflib\cpx\__init__.py", line 143, in run
    packet = self._transport.readPacket()
  File "d:\project\crazyflie-lib-python\cflib\cpx\transports.py", line 64, in readPacket
    size = struct.unpack('H', self._readData(2))[0]
  File "d:\project\crazyflie-lib-python\cflib\cpx\transports.py", line 60, in _readData
    data.extend(self._socket.recv(size-len(data)))
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host`

**Reply by knmcguire** [reply]
Hmmm... this is very strange. I wonder if your firewall has anything to do with this since I don't see anything like this on my windows 11

Do you by any chance have WSL2 on your computer? Or a linux dual boot

**Reply by noahzn** [reply]
It could be the firewall, but I cannot change the firewall setting because it's managed by our university. I just tried the code on my other Ubuntu18 laptop, and the video stream worked. Although I didn't make it fly, and since I'm using the RGB camera, the video has black pixels. I will try to replace the camera with a monochromatic camera and do a final test. I will close this discussion after that. Thank you very much for your help! @knmcguire, you already helped me a lot!

**Reply by noahzn** [reply]
It works well on my Ubuntu laptop with a grayscale camera. Thanks!

**Reply by knmcguire** [reply]
Great! 

Just mind if you do start flying, with older batteries it might be that there is too much going to the motor as it is trying to take of faster. That will be probably the cause you had before of it taking off and losing connection. What you could do is to figure out to reduce the speed of take-off or perhaps implement the [high level commander take off ](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/high_level_commander/#:~:text=def%20takeoff(self%2C%20absolute_height_m%2C%20duration_s%2C%20group_mask%3D0%2C%20yaw%3D0.0))at the beginning of the script as that has a smoother take off trajectory then just the very rough setpoint height.

**Reply by knmcguire** [reply]
Great! 

Just mind if you do start flying, with older batteries it might be that there is too much going to the motor as it is trying to take of faster. That will be probably the cause you had before of it taking off and losing connection. What you could do is to figure out to reduce the speed of take-off or perhaps implement the [high level commander take off ](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/high_level_commander/#:~:text=def%20takeoff(self%2C%20absolute_height_m%2C%20duration_s%2C%20group_mask%3D0%2C%20yaw%3D0.0))at the beginning of the script as that has a smoother take off trajectory then just the very rough setpoint height.

**Reply by noahzn** [reply]
OK. I think I have an old battery because we bought this crazyflie last year. 
So you suggest that I use `takeoff()` and maybe set the `duration_s` of taking off, instead of using `send_hover_setpoint()` in the fpv.py

**Reply by knmcguire** [reply]
yes indeed, I haven't tried it out myself but perhaps in the init function of the python script. Hopefully that works.
